internal/sync/cache.go

```diff
@@ -0,0 +1,362 @@
+	ignoreResourceUpdatesEnabled bool
+}
+
+type ObjectUpdatedHandler = func(managedByGitSync map[string]bool, ref v1.ObjectReference)
+
+type liveStateCache struct {
+	clusterCacheConfig *rest.Config
+	logger             *zap.SugaredLogger
+	onObjectUpdated    ObjectUpdatedHandler
+
+	cluster       clustercache.ClusterCache
+	cacheSettings cacheSettings
+	lock          sync.RWMutex
+}
+
+func NewLiveStateCache(
+	onObjectUpdated ObjectUpdatedHandler,
+) LiveStateCache {
+	return &liveStateCache{
+		onObjectUpdated: onObjectUpdated,
+	}
+}
+
+func (c *liveStateCache) getCluster() (clustercache.ClusterCache, error) {
+	c.lock.RLock()
+	cacheSettings := c.cacheSettings
+	c.lock.RUnlock()
+
+	if c.cluster != nil {
+		return c.cluster, nil
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.cluster != nil {
+		return c.cluster, nil
+	}
+
+	clusterCacheConfig := c.clusterCacheConfig
+
+	// TODO: check if we need set managed namespace,
+	//       also enable logging within clusterCache
+	clusterCacheOpts := []clustercache.UpdateSettingsFunc{
+		clustercache.SetListSemaphore(semaphore.NewWeighted(clusterCacheListSemaphoreSize)),
+		clustercache.SetListPageSize(clusterCacheListPageSize),
+		clustercache.SetListPageBufferSize(clusterCacheListPageBufferSize),
+		clustercache.SetWatchResyncTimeout(clusterCacheWatchResyncDuration),
+		clustercache.SetClusterSyncRetryTimeout(clusterSyncRetryTimeoutDuration),
+		clustercache.SetResyncTimeout(clusterCacheResyncDuration),
+		clustercache.SetSettings(cacheSettings.clusterSettings),
+		clustercache.SetPopulateResourceInfoHandler(c.PopulateResourceInfo),
+		clustercache.SetRetryOptions(clusterCacheAttemptLimit, clusterCacheRetryUseBackoff, isRetryableError),
+	}
+
+	clusterCache := clustercache.NewClusterCache(clusterCacheConfig, clusterCacheOpts...)
+
+	_ = clusterCache.OnResourceUpdated(func(newRes *clustercache.Resource, oldRes *clustercache.Resource, namespaceResources map[kube.ResourceKey]*clustercache.Resource) {
+		c.lock.RLock()
+		cacheSettings := c.cacheSettings
+		c.lock.RUnlock()
+
+		if cacheSettings.ignoreResourceUpdatesEnabled && oldRes != nil && newRes != nil && skipResourceUpdate(resInfo(oldRes), resInfo(newRes)) {
+			return
+		}
+
+		// TODO: when switch to change event triggering for sync tasking, notify
+		//       the GitSync managing the resources that there are changes happen
+	})
+
+	c.cluster = clusterCache
+
+	return clusterCache, nil
+}
+
+func (c *liveStateCache) PopulateResourceInfo(un *unstructured.Unstructured, isRoot bool) (interface{}, bool) {
+	res := &ResourceInfo{}
+	c.lock.RLock()
+	settings := c.cacheSettings
+	c.lock.RUnlock()
+
+	res.Health, _ = health.GetResourceHealth(un, settings.clusterSettings.ResourceHealthOverride)
+
+	gitSyncName := getGitSyncName(un)
+	if isRoot && gitSyncName != "" {
+		res.GitSyncName = gitSyncName
+	}
+
+	gvk := un.GroupVersionKind()
+
+	if settings.ignoreResourceUpdatesEnabled && shouldHashManifest("", gvk) {
```

Comment:
I notice ArgoCD passes in the Application name - why don't we pass in the GitSync name?